### PR TITLE
Synonyms with lower case

### LINKED ISSUE #576: Synonyms case sensitive (they shouldn't be)
We don't seem to normalize synonyms before they are inserted into the trie:

```
So, this doesn't work:

127.0.0.1:6379> ft.create emp schema name text
OK
127.0.0.1:6379> ft.synadd emp Alexander Alex
(integer) 0
127.0.0.1:6379> ft.add emp e11 1 fields name "Alexander"
OK
127.0.0.1:6379> ft.search emp alex
1) (integer) 0
127.0.0.1:6379> 

But this does, where I change the synadd to use lower case terms:

127.0.0.1:6379> ft.create emp schema name text
OK
127.0.0.1:6379> ft.synadd emp alexander alex
(integer) 0
127.0.0.1:6379> ft.add emp e11 1 fields name "Alexander"
OK
127.0.0.1:6379> ft.search emp alex
1) (integer) 1
2) "e11"
3) 1) "name"
   2) "Alexander"
```

### LINKED ISSUE #1196: Lowercase only?
Do synonyms support only lowercase?

This is OK:
"FT.CREATE" "my_idx" "SCHEMA" "my_fld" "TEXT" 
"FT.SYNADD" "my_idx" "hello" "hola" "hej" "bonjour"
"FT.ADD" "my_idx" "42" "1.0" "REPLACE" "FIELDS" "my_fld" "hello" 
"FT.SEARCH" "my_idx" "hej"

This is NOT OK:
"FT.CREATE" "my_idx2" "SCHEMA" "my_fld2" "TEXT" 
"FT.SYNADD" "my_idx2" "HELLO" "hola" "hej" "bonjour"
"FT.ADD" "my_idx2" "666" "1.0" "REPLACE" "FIELDS" "my_fld2" "HELLO" 
"FT.SEARCH" "my_idx2" "hej"

Why synonyms are not working with uppercase?

### PR DESCRIPTION
**Pull request by ashtul**
fix #1196
fix #576

### CODE REVIEW

File: src/synonym_map.c [hunk 1]
```diff
@@ -92,10 +92,12 @@ void SynonymMap_Free(SynonymMap* smap) {
   rm_free(smap);
 }
 
+char *strtolower(char *str); // in spec.c
```
Comment (MeirShpilraien) [inline src/synonym_map.c:95]: put in common.h

Comment (MeirShpilraien) [reply src/synonym_map.c:95]: Sorry common.h is not a good place, we need to have some common file to have all those functions. but till then why not put it in spec.h and include it?

Comment (ashtul) [reply src/synonym_map.c]: Is it OK that I'm changing the original RedisString or should I duplicate?

File: src/util/strconv.h [hunk 2]
```diff
@@ -56,4 +57,13 @@ static int ParseBoolean(const char *arg, int *res) {
   return 0;
 }
 
+static char *strtolower(char *str) {
```
Comment (MeirShpilraien) [inline src/util/strconv.h:60]: So its static and you are using it from outside? I do not think you are using the right one ..

Comment (MeirShpilraien) [reply src/util/strconv.h:60]: Also its not ok that you are changing the char* return from the RedisModuleString, there is a reason why it const.

File: src/synonym_map.c [hunk 3]
```diff
@@ -127,13 +127,15 @@ uint32_t SynonymMap_Add(SynonymMap* smap, const char** synonyms, size_t size) {
 
 void SynonymMap_Update(SynonymMap* smap, const char** synonyms, size_t size, uint32_t id) {
   RS_LOG_ASSERT(!smap->is_read_only, "SynonymMap should not be read only");
+  char *syn_str[size];
   int ret;
   for (size_t i = 0; i < size; i++) {
+    syn_str[i] = strtolower(rm_strdup(synonyms[i]));
```
Comment (MeirShpilraien) [inline src/synonym_map.c:133]: Why not just define it as a temporary char inside the for?

File: src/util/strconv.h (hunk 2 above)
Comment (MeirShpilraien) [reply src/util/strconv.h:60]: Still static? not follow how come you use it from outside?

Comment (ashtul) [reply src/util/strconv.h:60]: It is a header file which I include directly

File: src/synonym_map.c (hunk 3 above)
Comment (ashtul) [reply src/synonym_map.c:133]: @MeirShpilraien 
ASAN detects a memory leak here that was exposed once the `strdup` function was moved from within TermData_New in `line 136`.
Freeing when index is dropped seems correct. Still investigating.

Comment (ashtul) [reply src/synonym_map.c]: fixed

**Comment by ashtul** [comment]
> @ashtul this looks good in terms of code but now I notice that you basically turn synonyms to be case insensitive, I think this needs @K-Jo approval.

@MeirShpilraien 
The other way around. It used to be case sensitive.

**Comment by K-Jo** [comment]
> > @ashtul this looks good in terms of code but now I notice that you basically turn synonyms to be case insensitive, I think this needs @K-Jo approval.
> 
> @MeirShpilraien
> The other way around. It used to be case sensitive.

@ashtul is this for the 2.0?  what happens if I upgrade, will the trie still be case sensitive?

**Comment by ashtul** [comment]
@K-Jo 
At the moment it is not case sensitive. It is broken and will find neither lower or upper case.
I would cherry-pick for 1.6 as well and maybe even 1.4

**Comment by MeirShpilraien** [comment]
@ashtul the fact that it is not case sensitive does not mean that its broken. You are changing behavior here and this is why I asked for Pieter approval. And @K-Jo regarding your question it will not "fix" already existing indexes (on upgrade for example)

**Comment by K-Jo** [comment]
@ashtul @MeirShpilraien let's add it to the 2.0 as a breaking change.  And we need to add it to the next releases of 1.4 and 1.6 as a known issue for now.

**Comment by ashtul** [comment]
fix #1264

**Comment by massyah** [comment]
@ashtul it seems to me that stop words are also case sensitive on the current master. Do you think your branch would fix this also ? If not I'll open an appropriate issue with details

**Comment by ashtul** [comment]
fix #1264

**Comment by ashtul** [comment]
@massyah 
The fix would work for 1.6 but we are concern it is a breaking change so it will be added only to RS 2.0

File: src/util/strconv.h (hunk 2 above)
Comment (rafie) [inline src/util/strconv.h:60]: Note that this is only good for ANSI. When we have the means to handle Unicode (i.e. C++) we should handle this a little bit differently. See https://stackoverflow.com/questions/34433380/lowercase-of-unicode-character.
We can drop a TODO comment, just so we won't forget.

**Review by rafie — APPROVED** [review]

File: src/util/strconv.h (hunk 2 above)
Comment (ashtul) [reply src/util/strconv.h:60]: @rafie this deserves a separate issue

**Pull request merged**